### ISSUE TITLE
IPFS/IPLD/IPNS for versioning, dependency managing and invalidating content

### ISSUE THREAD

**Issue by venturecommunism** [apollostack/apollo#53]
I believe an essentially correct counterantidisintermediationist solution has been found to transport agnosticism on the issue @mitar raised. https://github.com/apollostack/apollo/issues/17

Deps would then be named merkle-links. JSON++ graphs of merkle-links form a DAG suited to dependency management and version control iff there are no SHA collisions.

Mutations fully determined on the client or server need only inform the server or client of the multihash. End-to-end, p2p.

With lexing and parsing and marshalling done by the IPFS ecosystem we can dispense with most of GraphQL except for GraphiQL and all the server specific code.

Probably something like this was the right way to build the entire file structure in POSIX the whole time.

Maybe I'm wrong? (I've almost definitely got some of the terminology from IPFS wrong, but that's another issue)

Can anyone pop the @MasterJames big ideas bubble?

https://github.com/ipfs/specs/tree/master/ipld
https://github.com/ipfs/js-ipld

**Comment by stubailo** [comment]
We are not currently interested in exploring dramatically new technologies for data transfer and synchronization, but rather in making more incremental improvements that address concrete problems people are experiencing with app development today. Big ideas are great, but I don't think IPFS is going to happen in the short term unless someone else decides to build it, or there is some very concrete reason people need this in their application stack.

**Comment by MasterJames** [comment]
Ya arguments are always possible. The point is to get people to think. and think clearer about the target and goals. Studying the history is how to get a bearing. 
Required Study do not miss out on the wisdom of the elders
https://www.youtube.com/watch?v=YyIQKBzIuBY
References 
https://www.computer.org/csdl/proceedings/afips/1961/5058/00/50580393.pdf
Ya it's dated (history?) so it's about how you incorporate it into your vision.

Meaning lets not completely loose target on the long term goals, which should be the simplest most efficient system design. I only muse here in hopes the right minds in the right position will adjust tack for the heart of our mutual human condition, when the time is right. 

I did not like the immutability of merkel dag in the IPFS form after about a week of wishful thinking.
I seem to still be gravitating to the AWS' EFS (Elastic File System) stuck in perpetual "Preview".
It does have a limit of 87 locks per file. I think it's manageable through Client requeueing or some similar mechanism like 87 masters per locked record (we'll cross that bridge if it's needed).
Here's the Master Plan so far hidden in a few poetic words to be decoded in deep thought.

```
**The Arrow of Time’s Master Plan in Poetic Form**

The Shared File System is a quorum with permissions and locking,
For Similar Citizen’s Self Mounted Containers are happily docking and talking.
They use EFS as an ETCD or Redis might otherwise be,
but fear not it’s all just trying to be me! EFS Simple. You need to see.

BJSON maybe NOT needed,
just deep JavaScript Objects with a referencing on focus was heeded.
The watched files notified the subscriber’s websocket, 
as their data changed, callbacks turned to hocus pocus... not needed
and feel to the wayside as the answers had already slipped into their desired pocket.

To decentralize? is a bigger fish for the deep dive fryers of Derivers to clearly see,
but it’s safe to say obfuscation links are the relation,
with random hash-ring paired job dissemination.
That makes it fast and secure without an e-coin to be seen.

Essentially the reproducible state of the Typeless JS running machine,
is bound to the irrelivant database within EFS “how obscene”,
Merging files, with code, and data with sockets and back to EFS, is not the last bean!

It was THREE.js 3D that laughed at the DOM, which was by then very DOM indeed, 
“But MJ it's just the canvas I see?! Nothing needed but EFS’ tree?”,
“That a is a right my son just let it be, virtually free. Now you're starting to see.”
the dag of code changes when signed off legit,
was the way to the micro-payments that was the fairest to permit.

The 3D Environment can be a bowlful of nodal graphs,
The code was simple and self building from visual flagstaffs.
When all of the code needed was written down once,
no more was needed and now was just a munce.
It took on a new form of visual cues,
that any old body could play with and muse.
Scientifically accurate simulated design,
a child could easily build that new printed house of mine.

A Guinness for the Genius who’s compiler-compiler makes it the smallest with “applause”
“Not possible!” he said as he squaked, thawed, and briefly paused.
Bloatware and code count to you I discount!
Staying out of my domain is a main rule and tantamount.

As the Human Lemmings ran head on straight on into the singularity,
it was those at the outer edges that found chrysotherapy,
which lead to the viviparous angularity and spared them the futile hilarity.

When the comp-clock did start,
the view was clear as the most beautiful art,
but was lost through the windows thinned of a few downwind,
who only chagrined when their mistake was merely skimmed.

 = = = = = 

www.master-domain.com lived for decades in a perpetual blackbox state,
waiting for a time when clarity and progress did inflate to be able to gestate.
Now we can see the cardboard world within,
when the “world within” can see us? Then we will be ready to begin.
Can you get into the blackbox? Of course you can. You are already there.
The journey is within, the question is can or should we open it? Maybe only together we can?


```


So far I can only speculate that AWS must see the devastating transformation that EFS or similar will bring. Will they open it? Will they be the first? Management must have serious control freak issues “Preview” for a year?! Somethings not right there.
So ya maybe skipping the hybrid state to decentralization will be needed via IPFS with a better multiple synchronized state machine layer as so eloquently encoded in the biblical poetic form it deserves above.

Basically if you take anything from this the lesson is the singularity is impossible and will flatten out with a hybrid of a complex but simple form. It all must be Crystallize into a dense but transparent form.

May the spirits of history and invention guide you to a better way (and brighter light).

**Comment by stubailo** [comment]
I think it's great to have some discussions about IPFS, but I don't think issues on this repository are the right place.

**Comment by MasterJames** [comment]
I actually completely agree. You guys are doing an awesome critical role here and doing it well. I thank you for that.
I leave you now (or the more interested or entertained viewer with enough time and patience) with some more succinct words of wisdom from our founders like Alan Kay.
https://www.youtube.com/watch?v=NdSD07U5uBs
